substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/JfrChunkWriter.java

```diff
@@ -371,43 +412,61 @@ private static int makePaddedInt(long sizeWritten) {
+        protected JfrCloseFileOperation() {
             // Some of the JDK code that deals with files uses Java synchronization. So, we need to
             // allow Java synchronization for this VM operation.
             super("JFR close file", SystemEffect.SAFEPOINT, true);
-            this.metadataDescriptor = metadataDescriptor;
-            this.repositories = repositories;
         }
 
         @Override
         protected void operate() {
-            changeEpoch();
-            try {
-                long constantPoolPosition = writeCheckpointEvent(repositories);
-                long metadataPosition = writeMetadataEvent(metadataDescriptor);
-                patchFileHeader(constantPoolPosition, metadataPosition);
-                file.close();
-            } catch (IOException e) {
-                Logger.log(JFR_SYSTEM, ERROR, "Error while writing file " + filename + ": " + e.getMessage());
+            boolean shouldNotify = changeEpoch();
+            if (shouldNotify) {
+                //Checkstyle: stop
+                synchronized (Target_jdk_jfr_internal_JVM.FILE_DELTA_CHANGE) {
+                    Target_jdk_jfr_internal_JVM.FILE_DELTA_CHANGE.notifyAll();
+                }
+                //Checkstyle: resume
             }
-
-            filename = null;
-            file = null;
         }
 
         @Uninterruptible(reason = "Prevent pollution of the current thread's thread local JFR buffer.")
-        private void changeEpoch() {
+        private boolean changeEpoch() {
             // TODO: We need to ensure that all JFR events that are triggered by the current thread
             // are recorded for the next epoch. Otherwise, those JFR events could pollute the data
             // that we currently try to persist. To ensure that, we must execute the following steps
             // uninterruptibly:
             //
-            // - Flush all thread-local buffers (native & Java) to global JFR memory.
+            // - Flush all buffers (native, Java and global) to disk
             // - Set all Java EventWriter.notified values
-            // - Change the epoch.
+            // - Change the epoch
+            boolean shouldNotify = false;
+            for (IsolateThread thread = VMThreads.firstThread(); thread.isNonNull(); thread = VMThreads.nextThread(thread)) {
+                JfrBuffer b = JfrThreadLocal.getJavaBuffer(thread);
+                if (b.isNonNull()) {
+                    writeAtSafepoint(b);
+                    JfrThreadLocal.notifyEventWriter(thread);
+                }
+                b = JfrThreadLocal.getNativeBuffer(thread);
+                if (b.isNonNull()) {
+                    writeAtSafepoint(b);
+                }
+            }
+
+            JfrBuffers buffers = globalMemory.getBuffers();
+            for (int i = 0; i < globalMemory.getBufferCount(); i++) {
+                JfrBuffer buffer = buffers.addressOf(i).read();
+                if (JfrBufferAccess.acquire(buffer)) {
```

Comment:
None of the global buffers may be acquired by another thread (otherwise, this would cause major issues as that might not get written into the correct chunk). So, it should be fine to replace `if (JfrBufferAccess.acquire(buffer))` with an assertion that the buffer is not acquired by any thread (and because we are at a safepoint in uninterruptible code, it is also unnecessary to acquire the buffer here).

Reply:
I've moved this out as suggested using a boolean field in the class.